### PR TITLE
added Axo

### PR DESCRIPTION
**Pull request by alexanderwatanabe**
Axo has tokens locked in many dev, ISPO and presale vesting contracts:

- dev 2.1m
- ISPO 1.05 (only ~50% were claimed of the 2.1m allocation)
- presale 2.1m

The addresses in `lockedRaw` are the largest vesting contracts sorted descending by tokens. If it is too many to include in the request, feel free to truncate after N addresses.

### CODE REVIEW

File: src/tokens/axo.ts
Comment (lloydswale) [inline src/tokens/axo.ts:1]: Too many addresses are causing the test to timeout

File: src/index.ts [hunk 1]
```diff
@@ -105,6 +106,7 @@ export const supplyFetchers: Record<string, SupplyFetcher> = {
   "577f0b1342f8f8f4aed3388b80a8535812950c7a892495c0ecdf0f1e0014df10464c4454":
     fldtFetcher,
   f43a62fdc3965df486de8a0d32fe800963589c41b38946602a0dc53541474958: agixFetcher,
+  "420000029ad9527271b1b1e3c27ee065c18df70a4a4cfc3093a41a44": axoFetcher,
```
Comment (lloydswale) [inline src/index.ts:109]: `420000029ad9527271b1b1e3c27ee065c18df70a4a4cfc3093a41a4441584f`

Comment (alexanderwatanabe) [reply src/index.ts]: thank you

File: src/tokens/axo.ts
Comment (alexanderwatanabe) [reply src/tokens/axo.ts:1]: kept top 100 addresses (list was sorted descending by value locked, addresses with <10,000 AXO were cut

File: src/tokens/axo.ts [hunk 2]
```diff
@@ -0,0 +1,286 @@
+import { defaultFetcherOptions, SupplyFetcher } from "../types";
+import { getAmountInAddresses, getBlockFrostInstance } from "../utils";
+
+const AXO = "420000029ad9527271b1b1e3c27ee065c18df70a4a4cfc3093a41a44";
```
[the remark this answers is not in the document]
Comment (alexanderwatanabe) [reply src/tokens/axo.ts]: thank you again

**Review by alexanderwatanabe — COMMENTED** [review]
incorporated @lloydswale's fixes

**Review by lloydswale — APPROVED** [review]
Nice work, shame about the missing addresses but at least its a decent representation of the circulating supply

**Review by lloydswale — APPROVED** [review]

**Review by lloydswale — CHANGES_REQUESTED** [review]
Test is giving out incorrect values for circulating and total supply
Screenshot 2024-01-18 at 07 18 38
If you make the changes as per my comments the new totals are:
Screenshot 2024-01-18 at 07 20 33

File: src/tokens/axo.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
+import { defaultFetcherOptions, SupplyFetcher } from "../types";
+import { getAmountInAddresses, getBlockFrostInstance } from "../utils";
+
+const AXO = "420000029ad9527271b1b1e3c27ee065c18df70a4a4cfc3093a41a4441584f";
+
+const fetcher: SupplyFetcher = async (options = defaultFetcherOptions) => {
+  const blockFrost = getBlockFrostInstance(options);
+  const total = 42_000_000e9; // 42 million
```
Comment (lloydswale) [inline src/tokens/axo.ts:8]: 42_000_000

File: src/tokens/axo.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    "addr1wyc3ka3enn04x0nm23jaxztvx0e4w357m5z7h59my49sduqznfw5j",
+    "addr1w8jrshxc56u7qmh48hrlx2n6pvdy39mprz70qsufheadrnc4t3hfs",
+    "addr1w86q8zfuzf73xfawmsfs8sdr9zy7ksjthqzpgrw7srrt2xqf8g6yh",
+    "addr1wyt0tfyghw5dqhauwy3gpqdaqed027drq4psn22km0ns9asgr8rts",
+    "addr1w93nnt0rcpr72a44c90ytqkgke46w22nx9stkhtvvcqep0spnpdfn",
+    "addr1w9cfscja9cunyxal7s2w4328uzly7fxtssl67mvvf28cdss7egh3m",
+    "addr1w8pj5nlnk3v89udlcph27jel2xxh049hswfcehq4vgjynkgvlvzzg",
+    "addr1wy0usqqcsdcwlayjz4aehvjny38cemnufspwyph9q54l7vg3972mx"
+  ]);
+  const locked = Number(lockedRaw) / 1e9;
```
Comment (lloydswale) [inline src/tokens/axo.ts:118]: 1_000_000_000

File: src/tokens/axo.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  const treasury = Number(treasuryRaw) / 1e9;
```
Comment (lloydswale) [inline src/tokens/axo.ts:119]: 1_000_000_000

**Review by lloydswale — APPROVED** [review]
Nice one brother! 
Happy to approve despite the inflated circulating supply due to the large no. of addresses causing the blockfrost API call to exceed 30s. 

Here's the current test output:
image

If anyone can think of a way to improve this that would be ideal

**Comment by alexanderwatanabe** [comment]
not the most elegant, how about this?

**Comment by lloydswale** [comment]
Nice work boss here's the new test output:
image

Is this consistent with the tokenomics?
image

**Comment by alexanderwatanabe** [comment]
i believe so, at launch when people claim their vesting amounts (20%) out of these contracts they should update.

i suppose this will slightly lag behind potential circulating tokens, as if people don't claim them often and leave them in the contracts, the supply will not update.

However, that's probably desirable behavior as if they aren't claimed after 18 months, they revert to treasury.

Thank you for the help Lloyd.

**Review by shadowkora — APPROVED** [review]

**Pull request merged**